From: IrfanMohammad <irfan@graphiant.com>
Date: Mon, 31 Jul 2023 15:54:59 +0000
Subject: TESTS sleep and retry with bigger step

 sleep of 10ms and retry is too aggressive, and the scheduler may
 not always get around scheduling the correct thread, when busy.
---
 tests/CMakeLists.txt       |  2 +-
 tests/test_apply_changes.c | 30 ++++++++++++++++--------------
 tests/test_lock_log.c      |  2 +-
 3 files changed, 18 insertions(+), 16 deletions(-)

diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
index a110777..db9002a 100644
--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -131,7 +131,7 @@ endif()
 # valgrind tests
 if(ENABLE_VALGRIND_TESTS)
     foreach(test_name IN LISTS tests)
-        add_test(NAME ${test_name}_valgrind COMMAND valgrind --malloc-fill=0xED --free-fill=0xBF --leak-check=full --show-leak-kinds=all --suppressions=${CMAKE_SOURCE_DIR}/tests/mongoc_init.supp --error-exitcode=1 ${CMAKE_BINARY_DIR}/tests/${test_name})
+        add_test(NAME ${test_name}_valgrind COMMAND valgrind --track-origins=yes --num-callers=24 --malloc-fill=0xED --free-fill=0xBF --leak-check=full --show-leak-kinds=all --suppressions=${CMAKE_SOURCE_DIR}/tests/mongoc_init.supp --error-exitcode=1 ${CMAKE_BINARY_DIR}/tests/${test_name})
         set(test_name "${test_name}_valgrind")
         set_property(TEST ${test_name} APPEND PROPERTY ENVIRONMENT
             "SYSREPO_REPOSITORY_PATH=${PROJECT_BINARY_DIR}/test_repositories/${test_name}"
diff --git a/tests/test_apply_changes.c b/tests/test_apply_changes.c
index ca99002..3cffd73 100644
--- a/tests/test_apply_changes.c
+++ b/tests/test_apply_changes.c
@@ -34,6 +34,9 @@
 #include "sysrepo.h"
 #include "tests/tcommon.h"
 
+#define RETRY_COUNT 500
+#define SLEEP_STEP_MS 100000
+
 struct state {
     sr_conn_ctx_t *conn;
     ATOMIC_T cb_called, cb_called2, cb_called3;
@@ -507,8 +510,8 @@ subscribe_change_done_thread(void *arg)
     pthread_barrier_wait(&st->barrier);
 
     count = 0;
-    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 1) && (count < 1500)) {
-        usleep(10000);
+    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 1) && (count < RETRY_COUNT)) {
+        usleep(SLEEP_STEP_MS);
         ++count;
     }
     assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 1);
@@ -4998,7 +5001,6 @@ subscribe_change_done_when_thread(void *arg)
 
     /* signal that subscription was created */
     pthread_barrier_wait(&st->barrier);
-
     /* wait for the other thread to finish */
     pthread_barrier_wait(&st->barrier);
 
@@ -5580,12 +5582,12 @@ apply_change_timeout_thread(void *arg)
     pthread_barrier_wait(&st->barrier);
 
     /* perform the change, time out but give it some time so that the callback is at least called) */
-    ret = sr_apply_changes(sess, 10);
+    ret = sr_apply_changes(sess, 100);
     assert_int_equal(ret, SR_ERR_CALLBACK_FAILED);
     pthread_barrier_wait(&st->barrier2);
 
     /* try again while the first callback is still executing (waiting) */
-    ret = sr_apply_changes(sess, 10);
+    ret = sr_apply_changes(sess, 100);
     assert_int_equal(ret, SR_ERR_CALLBACK_FAILED);
     pthread_barrier_wait(&st->barrier2);
 
@@ -5633,8 +5635,8 @@ subscribe_change_timeout_thread(void *arg)
     pthread_barrier_wait(&st->barrier);
 
     count = 0;
-    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 2) && (count < 1500)) {
-        usleep(10000);
+    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 2) && (count < RETRY_COUNT)) {
+        usleep(SLEEP_STEP_MS);
         ++count;
     }
     assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 2);
@@ -5643,8 +5645,8 @@ subscribe_change_timeout_thread(void *arg)
     pthread_barrier_wait(&st->barrier);
 
     count = 0;
-    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 4) && (count < 1500)) {
-        usleep(10000);
+    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 4) && (count < RETRY_COUNT)) {
+        usleep(SLEEP_STEP_MS);
         ++count;
     }
     assert_int_equal(ATOMIC_LOAD_RELAXED(st->cb_called), 4);
@@ -5731,7 +5733,7 @@ apply_done_timeout_thread(void *arg)
     /* try again, time out again (callback is called but gets stuck) */
     ret = sr_set_item_str(sess, "/test:test-leaf", "31", NULL, 0);
     assert_int_equal(ret, SR_ERR_OK);
-    ret = sr_apply_changes(sess, 100);
+    ret = sr_apply_changes(sess, 200);
     assert_int_equal(ret, SR_ERR_OK);
 
     /* unstuck the callback so it finishes */
@@ -5776,11 +5778,11 @@ subscribe_done_timeout_thread(void *arg)
 
     /* keep handling events */
     count = 0;
-    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 1) && (count < 1500)) {
+    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 1) && (count < RETRY_COUNT)) {
         ret = sr_subscription_process_events(subscr, NULL, NULL);
         assert_int_equal(ret, SR_ERR_OK);
 
-        usleep(10000);
+        usleep(SLEEP_STEP_MS);
         ++count;
     }
 
@@ -5789,11 +5791,11 @@ subscribe_done_timeout_thread(void *arg)
 
     /* keep handling events */
     count = 0;
-    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 2) && (count < 1500)) {
+    while ((ATOMIC_LOAD_RELAXED(st->cb_called) < 2) && (count < RETRY_COUNT)) {
         ret = sr_subscription_process_events(subscr, NULL, NULL);
         assert_int_equal(ret, SR_ERR_OK);
 
-        usleep(10000);
+        usleep(SLEEP_STEP_MS);
         ++count;
     }
 
diff --git a/tests/test_lock_log.c b/tests/test_lock_log.c
index 4e75154..40725d0 100644
--- a/tests/test_lock_log.c
+++ b/tests/test_lock_log.c
@@ -318,7 +318,7 @@ test_lock_dead_reader_helper(void)
 
     /* if we didn't crash in the callback,
      * because a lock blocked us from getting there, we will crash now */
-    assert_int_equal(ret, -1);
+    assert_int_equal(ret, 0xDECEA5E);
 
     sr_release_data(data);
     data = NULL;
